### DISCUSSION TITLE
Access data from Solar Assistant Broker using ESP8266 NodeMCU with usb port via console

### DISCUSSION THREAD

**Discussion by Daveolw** [arendst/Tasmota#19985] [Q&A]
Using tasmota console.
Trying to recieve data from Solar Assistant Broker, have set up tasmota as client but cannot get the correct command to make it work.  The topic is "solar_assistant/total/battery_state_of_charge/state" and the reply should be the value of Battery state of charge ie: 0-100
It needs subscribe so I added the:       
#ifndef SUPPORT_MQTT_EVENT
#define SUPPORT_MQTT_EVENT
#endif
to the user_config_override.h and compiled and uploaded the new bin file.
 Subscribe <eventName>, <mqttTopic> [, <key>]
Have tried subscribe with and without event name.
22:13:49.792 CMD: subscribe bat, cmnd/solar_assistant/total/battery_state_of_charge/state,
22:13:49.801 MQT: stat/tasmota_5FE66D/RESULT = 
{"Subscribe":"BAT,cmnd/solar_assistant/total/battery_state_of_charge/state/#"}
it adds /# do not know why?
But no return data.
Doc's show
Telemetry data will be sent by prefix tele like tele/tasmota/SENSOR {"Time":"2017-02-16T10:13:52", "DS18B20":{"Temperature":20.6}}
So have tried    "subscribe bat, tele/solar_assistant/total/battery_state_of_charge/state,"
it adds /# do not know why?
But no return data.
Under information it shows:
MQTT Host	192.168.10.42
MQTT Port	1883
MQTT User	DVES_USER
MQTT Client	makerio_mqtt
MQTT Topic	tasmota_%06X
MQTT Group Topic 1	cmnd/tasmotas/
MQTT Full Topic	cmnd/tasmota_5FE66D/
MQTT Fallback Topic	cmnd/makerio_mqtt_fb/
MQTT No Retain	Disabled
Why does it show under  MQTT Full Topic   "cmnd/tasmota_5FE66D" i thought the Full topic is the 
"solar_assistant/total/battery_state_of_charge/state" that i keep sending.
Have tried the following with a parameter at the end is that where it returns data?
subscribe bat, cmnd/tasmota_5FE66D/solar_assistant/total/battery_state_of_charge/state, var1
subscribe cmnd/tasmota_5FE66D/solar_assistant/total/battery_state_of_charge/state
subscribe bat, cmnd/tasmota_5FE66D/solar_assistant/total/battery_state_of_charge/state ?
subscribe cmnd/tasmota_5FE66D/solar_assistant/total/battery_state_of_charge/state -n

Where am I going wrong what should i be sending to get the data back?

Keeps coming up with these entries:
10:31:22.377 MQT: tele/tasmota_5FE66D/STATE = {"Time..........
10:26:12.379 APP: Serial logging disabled

Thanks

**Comment by sfromis** [comment]
Use command `Subscribe2` instead of `Subscribe` to avoid the confusion around `/#` added. If you intend to subscribe to data from another tasmota device, better not use the `cmnd` prefix for non-command data, but you need to be sure of exactly what is actually published, what topics are being used by the other device.

The device from which you are showing `MQTT Full Topic cmnd/tasmota_5FE66D/` does not have a FullTopic configured as you seem to expect. Is that the device from which the data is expected to come, or is it the device where you are doing the Subscribe? In the latter case, it is not related to subscribe picking up data or not.

And when using Subscribe, you absolutely need rules to pick up the events generated by the subscribe feature. You are not showing if or what you did there.

**Reply by Daveolw** [reply]
Thanks for the reply.
Not sending to another tasmota device, sending to Solar Assistant Broker on my local network (a raspberry pi)
so use "subscribe2 bat, solar_assistant/total/battery_state_of_charge/state, var1"
bat = eventname
solar_assistant/total/battery_state_of_charge/state = topic
var1 = parmeter
The "tasmota_5FE66D" is my ESP8266 which i am sending Subscribe from.
Had a look at "rule docs" bit confusing.
so:-      rule1 ON bat DO solar_assistant/total/battery_state_of_charge/state#?#var1
Is that correct?
sending that I get:-
12:58:42.840 MQT: stat/tasmota_5FE66D/RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":68,"Free":443,"Rules":"ON bat DO solar_assistant/total/battery_state_of_charge/state#?#var1"}}
Then sending:-
"subscribe2 bat, solar_assistant/total/battery_state_of_charge/state, var1"
Do not get a value.
Thanks

**Reply by sfromis** [reply]
Your "rule" and subscribe command looks like only a **very** cursory look at the docs. Take a look at the examples in https://tasmota.github.io/docs/MQTT/#subscribeunsubscribe

**Reply by Daveolw** [reply]
Changed the rule to:-
Rule1  ON mqtt#connected DO Subscribe2 bat, solar_assistant/total/battery_state_of_charge/state ENDON   ON Event#bat=ON DO var1 ENDON
Still not working.

**Reply by sfromis** [reply]
Did you also enable the rule set and restart?

In general, you are just wasting time posting "not working" without giving details.

**Comment by barbudor** [comment]
SolarAssistant seems to come with it's own MQTT broker or can use an external one such as one provided with HA (described as Advanced Setup i the docs)

In which case are you ? Using SA's own broker ?
Is you Tasmota device connected to the same MQTT broker as SA ?
Are you sure not to have multiple MQTT broker running and conflicting ?

Basic local test: If you have ran the command `publish solar_assistant/total/battery_state_of_charge/state ON` on your tasmota console, does anything happens ? what do you seen in logs ?

**Reply by Daveolw** [reply]
Yes using and connected to SA's own broker, it is on a Raspberry pi installed on Hybrid Inverter and connected to my local network.
 ran the command publish solar_assistant/total/battery_state_of_charge/state ON on tasmota console, 
I get
16:44:51.117 CMD: publish solar_assistant/total/battery_state_of_charge/state ON
16:44:51.124 MQT: solar_assistant/total/battery_state_of_charge/state = ON

**Reply by sfromis** [reply]
Well, if you have a subscribe to match that, you should see your rule triggering.
